esphome/components/atm90e32/atm90e32.cpp

```diff
@@ -0,0 +1,167 @@
+}
+
+void ATM90E32Component::dump_config() {
+  ESP_LOGCONFIG("", "ATM90E32:");
+  LOG_PIN("  CS Pin: ", this->cs_);
+  if (this->is_failed()) {
+    ESP_LOGE(TAG, "Communication with ATM90E32 failed!");
+  }
+  LOG_UPDATE_INTERVAL(this);
+  LOG_SENSOR("  ", "VoltageA", this->voltage_sensor_a_)
+  LOG_SENSOR("  ", "VoltageB", this->voltage_sensor_b_)
+  LOG_SENSOR("  ", "CurrentA", this->current_sensor_a_)
+  LOG_SENSOR("  ", "CurrentB", this->current_sensor_b_)
+  LOG_SENSOR("  ", "CurrentC", this->current_sensor_c_)
+  LOG_SENSOR("  ", "Frequency", this->freq_sensor_)
+  LOG_SENSOR("  ", "Power", this->power_sensor_)
+}
+float ATM90E32Component::get_setup_priority() const { return setup_priority::DATA; }
+
+uint16_t ATM90E32Component::read16_(uint16_t a_register) {
+  uint8_t addrh = (1 << 7) | ((a_register >> 8) & 0x03);
+  uint8_t addrl = (a_register & 0xFF);
+  uint8_t data[2];
+  uint16_t output;
+
+  this->enable();
+  delayMicroseconds(10);
+  this->write_byte(addrh);
+  this->write_byte(addrl);
+  delayMicroseconds(4);
+  this->read_array(data, 2);
+  this->disable();
+
+  output = (uint16_t(data[0] & 0xFF) << 8) | (data[1] & 0xFF);
+  ESP_LOGVV(TAG, "read16_ 0x%04X output 0x%04X", a_register, output);
+  return output;
+}
+
+uint32_t ATM90E32Component::read32_(uint16_t addr_h, uint16_t addr_l) {
+  uint32_t val_h = this->read16_(addr_h);
+  uint32_t val_l = this->read16_(addr_l);
+
+  return (val_h << 16) | val_l;
+}
+
+void ATM90E32Component::write16_(uint16_t a_register, uint16_t val) {
+  uint8_t addrh = (a_register >> 8) & 0x03;
+  uint8_t addrl = (a_register & 0xFF);
+
+  ESP_LOGVV(TAG, "write16_ 0x%04X val 0x%04X", a_register, val);
+  this->enable();
+  delayMicroseconds(10);
+  this->write_byte(addrh);
+  this->write_byte(addrl);
+  delayMicroseconds(4);
+  this->write_byte((val >> 8) & 0xff);
+  this->write_byte(val & 0xFF);
+  this->disable();
+}
+
+float ATM90E32Component::GetLineVoltageA() {
+  uint16_t voltage = this->read16_(ATM90E32_REGISTER_URMSA);
+  return (float)voltage / 100;
+}
+float ATM90E32Component::GetLineVoltageB() {
+  uint16_t voltage = this->read16_(ATM90E32_REGISTER_URMSB);
+  return (float)voltage / 100;
+}
+float ATM90E32Component::GetLineVoltageC() {
+  uint16_t voltage = this->read16_(ATM90E32_REGISTER_URMSC);
+  return (float)voltage / 100;
+}
+float ATM90E32Component::GetLineCurrentA() {
+  uint16_t current = this->read16_(ATM90E32_REGISTER_IRMSA);
+  return (float)current / 1000;
+}
+float ATM90E32Component::GetLineCurrentB() {
+  uint16_t current = this->read16_(ATM90E32_REGISTER_IRMSB);
+  return (float)current / 1000;
+}
+float ATM90E32Component::GetLineCurrentC() {
+  uint16_t current = this->read16_(ATM90E32_REGISTER_IRMSC);
+  return (float)current / 1000;
+}
+float ATM90E32Component::GetTotalActivePower() {
+   int val = this->read32_(ATM90E32_REGISTER_PMEANT, ATM90E32_REGISTER_PMEANTLSB);
+   return (float)val * 0.00032;
```

Comment:
```suggestion
   return val * 0.00032f;
```

Shorter and doesn't promote to `double`